Add automatic handling of Declarative Web Push messages in the UIProcess

beidson opened the pull request:
#### e5b61a361d97c315d2a0ba17ae6d323798ba03d1
<pre>
Add automatic handling of Declarative Web Push messages in the UIProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=261694">https://bugs.webkit.org/show_bug.cgi?id=261694</a>
rdar://115675105

Reviewed by Andy Estes and Chris Dumez.

The goal of Declarative Web Push is to handle the push message automatically.

Eventually we&apos;ll be able to do so without even requiring the UIProcess to launch.
But for now, we can do so inside the UIProcess bypassing any JavaScript.

This patch implements that logic, and tests that both the declarative notification
was shown and that the declarative app badge was set.

* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::data const):
* Source/WebCore/Modules/notifications/NotificationData.cpp:
(WebCore::NotificationData::isolatedCopy const):
(WebCore::NotificationData::isolatedCopy):
* Source/WebCore/Modules/notifications/NotificationData.h:
* Source/WebCore/Modules/notifications/NotificationDataCocoa.mm:
(WebCore::NotificationData::fromDictionary):
(WebCore::NotificationData::dictionaryRepresentation const):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPushMessage.cpp: Added.
(WebKit::WebPushMessage::notificationPayloadToCoreData const):
* Source/WebKit/Shared/WebPushMessage.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APINotificationProvider.h:
(API::NotificationProvider::show):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _processPushMessage:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration declarativeWebPushEnabled]):
(-[_WKWebsiteDataStoreConfiguration setDeclarativeWebPushEnabled:]):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotification.cpp:
(WebKit::WebNotification::WebNotification):
* Source/WebKit/UIProcess/Notifications/WebNotification.h:
(WebKit::WebNotification::createNonPersistent):
(WebKit::WebNotification::createPersistent):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp:
(WebKit::WebNotificationManagerProxy::showPersistent):
(WebKit::WebNotificationManagerProxy::showImpl):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.cpp:
(WebKit::WebNotificationProvider::show):
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::showPersistentNotification):
(WebKit::WebsiteDataStore::processPushMessage):
(WebKit::WebsiteDataStore::showServiceWorkerNotification): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::declarativeWebPushEnabled const):
(WebKit::WebsiteDataStoreConfiguration::setDeclarativeWebPushEnabled):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/mac/AppDelegate.m:
(enabledForFeature):
(-[BrowserAppDelegate persistentDataStore]):
(-[BrowserAppDelegate defaultConfiguration]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(-[PushNotificationDelegate websiteDataStore:showNotification:]):
(-[PushNotificationDelegate websiteDataStore:workerOrigin:updatedAppBadge:]):
(-[PushNotificationDelegate websiteDataStore:showNotification:]):
(-[NotificationPermissionDelegate _webView:requestNotificationPermissionForSecurityOrigin:decisionHandler:]): Deleted.
(-[NotificationPermissionDelegate webView:runJavaScriptAlertPanelWithMessage:initiatedByFrame:completionHandler:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/268101@main">https://commits.webkit.org/268101@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a7ae54b998f7698d425c8d90548c7944d1f6b2f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19040 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19181 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18924 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19076 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16278 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21441 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16299 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17038 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23491 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/17324 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17210 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21385 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17814 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16873 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4442 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21240 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17653 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
```diff
@@ -109,6 +109,9 @@ void NetworkSessionCreationParameters::encode(IPC::Encoder& encoder) const
 #endif
     encoder << resourceLoadStatisticsParameters;
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    encoder << declarativeWebPushEnabled;
```
Missing prefix as per WebKit coding style. Should something like `isDeclarativeWebPushEnabled`.

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+    static NeverDestroyed<WebCore::ScriptExecutionContextIdentifier> sharedScriptIdentifier = WebCore::ScriptExecutionContextIdentifier::generate();
+
+    String body, iconURL, tag, language;
+    WebCore::NotificationDirection direction = WebCore::NotificationDirection::Auto;
```
could use `auto`

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+        sharedScriptIdentifier,
+        PAL::SessionID::defaultSessionID(),
+        MonotonicTime::now(),
+        dataJSON,
```
WTFMove(dataJSON)

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+    return {
+        notificationPayload->defaultActionURL,
+        notificationPayload->title,
+        body,
```
WTFMove()

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+        notificationPayload->defaultActionURL,
+        notificationPayload->title,
+        body,
+        iconURL,
```
ditto.

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+        notificationPayload->title,
+        body,
+        iconURL,
+        tag,
```
ditto

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+        body,
+        iconURL,
+        tag,
+        language,
```
ditto

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+    WebCore::NotificationDirection direction = WebCore::NotificationDirection::Auto;
+    std::optional<bool> silent;
+
+    CString dataCString;
```
We don't need this to have that wide a scope.

cdumez commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+        direction = notificationPayload->options->dir;
+        silent = notificationPayload->options->silent;
+
+        dataCString = notificationPayload->options->dataJSONString.utf8();
```
CString dataCString = ...

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
```diff
@@ -580,6 +580,23 @@ - (void)setVolumeCapacityOverride:(NSNumber *)mockVolumeCapactiy
+- (BOOL)declarativeWebPushEnabled
+{
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    return _configuration->declarativeWebPushEnabled();
```
Would be nice to use a `is` prefix.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
```diff
@@ -580,6 +580,23 @@ - (void)setVolumeCapacityOverride:(NSNumber *)mockVolumeCapactiy
+{
+    UNUSED_PARAM(enabled);
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    _configuration->setDeclarativeWebPushEnabled(enabled);
```
`setIsDeclarativeWebPushEnabled()`

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1953,6 +1953,10 @@ WebsiteDataStoreParameters WebsiteDataStore::parameters()
     networkSessionParameters.inspectionForServiceWorkersAllowed = m_inspectionForServiceWorkersAllowed;
 #endif
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    networkSessionParameters.declarativeWebPushEnabled = m_configuration->declarativeWebPushEnabled();
```
isDeclarativeWebPushEnabled

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -78,6 +78,11 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
     void setVolumeCapacityOverride(std::optional<uint64_t> capacity) { m_volumeCapacityOverride = capacity; }
 
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    bool declarativeWebPushEnabled() const { return m_declarativeWebPushEnabled; }
```
Would be nice to use the `is` prefix.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -313,6 +318,9 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
     bool m_allowsHSTSWithUntrustedRootCertificate { false };
     bool m_trackingPreventionDebugModeEnabled { false };
+#if ENABLE(DECLARATIVE_WEB_PUSH)
+    bool m_declarativeWebPushEnabled { false };
```
ditto.

aestes commented on Source/WebKit/Shared/WebPushMessage.cpp:
```diff
@@ -0,0 +1,81 @@
+
+namespace WebKit {
+
+#if ENABLE(DECLARATIVE_WEB_PUSH)
```
Nit: I was surprised to see this line here instead of on line 28.

aestes commented on Source/WebKit/UIProcess/Notifications/WebNotification.cpp:
```diff
@@ -31,7 +31,7 @@
 namespace WebKit {
 
-WebNotification::WebNotification(const WebCore::NotificationData& data, WebPageProxyIdentifier pageIdentifier, const std::optional<WTF::UUID>& dataStoreIdentifier, IPC::Connection& sourceConnection)
+WebNotification::WebNotification(const WebCore::NotificationData& data, WebPageProxyIdentifier pageIdentifier, const std::optional<WTF::UUID>& dataStoreIdentifier, IPC::Connection* sourceConnection)
```
I couldn't quite follow why IPC::Connection was changed from a reference to a pointer.

beidson replied:
There are now some notifications that don't come from *any* XPC service, so they have no source connection

cdumez reviewed (approved):
LGTM % nits

beidson committed:
Add automatic handling of Declarative Web Push messages in the UIProcess
https://bugs.webkit.org/show_bug.cgi?id=261694
rdar://115675105

Reviewed by Andy Estes and Chris Dumez.

The goal of Declarative Web Push is to handle the push message automatically.

Eventually we'll be able to do so without even requiring the UIProcess to launch.
But for now, we can do so inside the UIProcess bypassing any JavaScript.

This patch implements that logic, and tests that both the declarative notification
was shown and that the declarative app badge was set.

* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::data const):
* Source/WebCore/Modules/notifications/NotificationData.cpp:
(WebCore::NotificationData::isolatedCopy const):
(WebCore::NotificationData::isolatedCopy):
* Source/WebCore/Modules/notifications/NotificationData.h:
* Source/WebCore/Modules/notifications/NotificationDataCocoa.mm:
(WebCore::NotificationData::fromDictionary):
(WebCore::NotificationData::dictionaryRepresentation const):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPushMessage.cpp: Added.
(WebKit::WebPushMessage::notificationPayloadToCoreData const):
* Source/WebKit/Shared/WebPushMessage.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APINotificationProvider.h:
(API::NotificationProvider::show):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _processPushMessage:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration declarativeWebPushEnabled]):
(-[_WKWebsiteDataStoreConfiguration setDeclarativeWebPushEnabled:]):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotification.cpp:
(WebKit::WebNotification::WebNotification):
* Source/WebKit/UIProcess/Notifications/WebNotification.h:
(WebKit::WebNotification::createNonPersistent):
(WebKit::WebNotification::createPersistent):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp:
(WebKit::WebNotificationManagerProxy::showPersistent):
(WebKit::WebNotificationManagerProxy::showImpl):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.cpp:
(WebKit::WebNotificationProvider::show):
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::showPersistentNotification):
(WebKit::WebsiteDataStore::processPushMessage):
(WebKit::WebsiteDataStore::showServiceWorkerNotification): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::declarativeWebPushEnabled const):
(WebKit::WebsiteDataStoreConfiguration::setDeclarativeWebPushEnabled):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/MiniBrowser/mac/AppDelegate.m:
(enabledForFeature):
(-[BrowserAppDelegate persistentDataStore]):
(-[BrowserAppDelegate defaultConfiguration]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(-[PushNotificationDelegate websiteDataStore:showNotification:]):
(-[PushNotificationDelegate websiteDataStore:workerOrigin:updatedAppBadge:]):
(-[PushNotificationDelegate websiteDataStore:showNotification:]):
(-[NotificationPermissionDelegate _webView:requestNotificationPermissionForSecurityOrigin:decisionHandler:]): Deleted.
(-[NotificationPermissionDelegate webView:runJavaScriptAlertPanelWithMessage:initiatedByFrame:completionHandler:]): Deleted.

Canonical link: https://commits.webkit.org/268101@main
